File: python/monkey/ast.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from monkey.token import Token


@dataclass
class Node(ABC):
    token: Token

    def token_literal(self) -> str:
        return self.token.literal


@dataclass
class Statement(Node):
    @abstractmethod
    def statement_node(self) -> None:
        pass


@dataclass
class Expression(Node):
    @abstractmethod
    def expression_node(self) -> None:
        pass


class Program:
    def __init__(self) -> None:
        self.statements: list[Statement] = []

    def token_literal(self) -> str:
        if self.statements:
            return self.statements[0].token_literal()
        return ""


@dataclass
class Identifier(Expression):
    value: str

    def expression_node(self) -> None:
        pass


@dataclass
class IntegerLiteral(Expression):
    value: int = 0

    def expression_node(self) -> None:
        pass


@dataclass
class Boolean(Expression):
    value: bool

    def expression_node(self) -> None:
        pass


@dataclass
class BlockStatement(Expression):
    statements: list[Statement]

    def expression_node(self) -> None:
        pass


@dataclass
class IfExpression(Expression):
    condition: Expression = None
    consequence: BlockStatement = None
    Alternative: BlockStatement = None

    def expression_node(self) -> None:
        pass


@dataclass
class FunctionLiteral(Expression):
    parameters: list[Identifier]
    body: BlockStatement = None

    def expression_node(self) -> None:
        pass


@dataclass
class CallExpression(Expression):
    function: Expression
    arguments: list[Expression]

    def expression_node(self) -> None:
        pass


@dataclass
class PrefixExpression(Expression):
    operator: str
    right: Expression = None

    def expression_node(self) -> None:
        pass


@dataclass
class InfixExpression(Expression):
    left: Expression = None
    operator: str = ""
    right: Expression = None

    def expression_node(self) -> None:
        pass


@dataclass
class LetStatement(Statement):
    name: Identifier = None
    value: Expression = None

    def statement_node(self):
        pass


@dataclass
class ReturnStatement(Statement):
    return_value: Expression = None

    def statement_node(self) -> None:
        pass


@dataclass
class ExpressionStatement(Statement):
    expression: Expression = None

    def statement_node(self) -> None:
        pass

File: csharp2/Monkey/Ast.cs
using System.Text;

namespace Monkey;

interface INode {
    Token Token { get; set; }
}

interface IStatement : INode { }

interface IExpression : INode { }

class LetStatement : IStatement {
    public required Token Token { get; set; }
    public required Identifier Name { get; set; }
    public IExpression? Value { get; set; }

    public override string ToString() {
        var builder = new StringBuilder($"{Token.Literal} {Name.Token.Literal} = ");
        if (Value is not null)
            builder.Append(Value.ToString());

        builder.Append(';');
        return builder.ToString();
    }
}

class ReturnStatement : IStatement {
    public required Token Token { get; set; }
    public IExpression? Expression { get; set; }

    public override string ToString() {
        var builder = new StringBuilder($"{Token.Literal} ");
        if (Expression is not null)
            builder.Append(Expression.ToString());

        builder.Append(';');
        return builder.ToString();
    }
}

class ExpressionStatement : IStatement {
    public required Token Token { get; set; }
    public IExpression? Expression { get; set; }

    public override string ToString() => Expression?.ToString() ?? "";
}

class Identifier : IExpression {
    public required Token Token { get; set; }
    public required string Value { get; set; }

    public override string ToString() => Value;
}

class IntegerLiteral : IExpression {
    public required Token Token { get; set; }
    public required long Value { get; set; }

    public override string ToString() => Token.Literal;
}

class PrefixExpression : IExpression {
    public required Token Token { get; set; }
    public required string Operator { get; set; }
    public IExpression? Right { get; set; }

    public override string ToString() {
        var builder = new StringBuilder($"({Operator}");
        if (Right is not null)
            builder.Append(Right.ToString());

        builder.Append(')');
        return builder.ToString();
    }
}

class InfixExpression : IExpression {
    public required Token Token { get; set; }
    public required string Operator { get; set; }
    public IExpression? Left { get; set; }
    public IExpression? Right { get; set; }

    public override string ToString() {
        var builder = new StringBuilder("(");
        if (Left is not null)
            builder.Append(Left.ToString());

        builder.Append($" {Operator} ");
        if (Right is not null)
            builder.Append(Right.ToString());

        builder.Append(')');
        return builder.ToString();
    }
}

class BooleanExpression : IExpression {
    public required Token Token { get; set; }
    public required bool Value { get; set; }

    public override string ToString() => Token.Literal;
}

class Ast {
    public List<IStatement> Statements { get; private set; } = new();

    public Ast(IEnumerable<IStatement>? statements = null) {
        if (statements is not null)
            Statements.AddRange(statements);
    }

    public override string ToString() {
        var builder = new StringBuilder();

        foreach (var statement in Statements)
            builder.Append(statement.ToString());

        return builder.ToString();
    }
}
